src/Tasks/PopulateSupportedArchitectures.cs

```diff
@@ -0,0 +1,227 @@
+
+        /// <summary>
+        /// Represents the result of validating an application manifest.
+        /// </summary>
+        private enum ManifestValidationResult
+        {
+            /// <summary>
+            /// The manifest validation was successful.
+            /// </summary>
+            Success = 1,
+
+            /// <summary>
+            /// The manifest validation failed.
+            /// </summary>
+            Failure,
+
+            /// <summary>
+            /// The supported architectures exist in the manifest with the expected value.
+            /// </summary>
+            SupportedArchitecturesExists,
+        }
+
+        /// <summary>
+        /// Path to the existing application manifest.
+        /// </summary>
+        public string? ApplicationManifestPath { get; set; }
+
+        /// <summary>
+        /// Intermediate output directory.
+        /// </summary>
+        [Required]
+        public string OutputDirectory
+        {
+            get => _outputDirectory;
+            set => _outputDirectory = value ?? throw new ArgumentNullException(nameof(OutputDirectory));
+        }
+
+        /// <summary>
+        /// Value for supportedArchitectures node.
+        /// </summary>
+        [Required]
+        public string SupportedArchitectures
+        {
+            get => _supportedArchitectures;
+            set => _supportedArchitectures = value ?? throw new ArgumentNullException(nameof(SupportedArchitectures));
+        }
+
+        /// <summary>
+        /// Returns path to the generated manifest.
+        /// </summary>
+        [Output]
+        public string ManifestPath
+        {
+            get => _generatedManifestFullPath;
+            private set => _generatedManifestFullPath = value;
+        }
+
+        private string? GetPathToManifest()
+        {
+            if (!string.IsNullOrEmpty(ApplicationManifestPath))
+            {
+                if (!File.Exists(ApplicationManifestPath))
+                {
+                    Log.LogErrorWithCodeFromResources("PopulateSupportedArchitectures.SpecifiedApplicationManifestCanNotBeFound", ApplicationManifestPath);
+                    return null;
+                }
+
+                return ApplicationManifestPath;
+            }
+
+            // The logic for getting default manifest is similar to the one from Roslyn:
+            // If Roslyn logic returns null, we fall back to reading embedded manifest.
+            return ToolLocationHelper.GetPathToDotNetFrameworkFile(DefaultManifestName, TargetDotNetFrameworkVersion.Version46);
+        }
+
+        public override bool Execute()
+        {
+            string? pathToManifest = GetPathToManifest();
+
+            // Only if ApplicationManifest was not specified, we can try to load the embedded manifest.
+            if (!(string.IsNullOrEmpty(pathToManifest) || (string.IsNullOrEmpty(pathToManifest) && string.IsNullOrEmpty(ApplicationManifestPath))))
```

Comment:
nit: Isn't `if (!(A || (A && B)))` equivalent to `if (!(A))` ?

```suggestion
            if (!string.IsNullOrEmpty(pathToManifest))
```